src/Features/Core/Portable/Diagnostics/Analyzers/PreferFrameworkTypeDiagnosticAnalyzerBase.cs

```diff
@@ -70,45 +88,70 @@ protected void AnalyzeNode(SyntaxNodeAnalysisContext context)
+            // check the appropriate option and determine if we should report a diagnostic.
             if (IsInMemberAccessOrCrefReferenceContext(predefinedTypeNode))
             {
-                applicableOption = CodeStyleOptions.PreferIntrinsicPredefinedTypeKeywordInMemberAccess;
-                applicableDiagnosticId = IDEDiagnosticIds.PreferFrameworkTypeInMemberAccessDiagnosticId;
+                diagnosticId = IDEDiagnosticIds.PreferFrameworkTypeInMemberAccessDiagnosticId;
+                optionValue = optionSet.GetOption(GetOptionForMemberAccessContext, language);
             }
             else
             {
-                applicableOption = CodeStyleOptions.PreferIntrinsicPredefinedTypeKeywordInDeclaration;
-                applicableDiagnosticId = IDEDiagnosticIds.PreferFrameworkTypeInDeclarationsDiagnosticId;
+                diagnosticId = IDEDiagnosticIds.PreferFrameworkTypeInDeclarationsDiagnosticId;
+                optionValue = optionSet.GetOption(GetOptionForDeclarationContext, language);
             }
 
-            var optionValue = optionSet.GetOption(applicableOption, semanticModel.Language);
-            var preferFrameworkType = !optionValue.Value;
+            severity = optionValue.Notification.Value;
+            return OptionSettingPrefersFrameworkType(optionValue, severity);
+        }
 
-            if (preferFrameworkType && optionValue.Notification.Value != DiagnosticSeverity.Hidden)
-            {
-                var descriptor = new DiagnosticDescriptor(applicableDiagnosticId,
-                        s_preferFrameworkTypeTitle,
-                        s_preferFrameworkTypeMessage,
-                        DiagnosticCategory.Style,
-                        optionValue.Notification.Value,
-                        isEnabledByDefault: true);
+        private bool IsStylePreferred(OptionSet optionSet, string language) =>
+            IsFrameworkTypePreferred(optionSet, GetOptionForDeclarationContext, language) ||
+            IsFrameworkTypePreferred(optionSet, GetOptionForMemberAccessContext, language);
 
-                context.ReportDiagnostic(Diagnostic.Create(descriptor, predefinedTypeNode.GetLocation()));
-            }
+        private bool IsFrameworkTypePreferred(OptionSet optionSet, PerLanguageOption<CodeStyleOption<bool>> option, string language)
+        {
+            var optionValue = optionSet.GetOption(option, language);
+            return OptionSettingPrefersFrameworkType(optionValue, optionValue.Notification.Value);
         }
+
+        /// <summary>
+        /// checks if style is preferred and the enforcement is not None.
+        /// </summary>
+        /// <remarks>if predefined type is not preferred, it implies the preference is framework type.</remarks>
+        private static bool OptionSettingPrefersFrameworkType(CodeStyleOption<bool> optionValue, DiagnosticSeverity severity) =>
+            !optionValue.Value && severity != DiagnosticSeverity.Hidden;
```

Comment:
I would make this `<returns>True if the predefined type keyword is preferred, false if the framework type name is preferred</returns>`